MariaDB/server
# MDEV-33974 Enable GNU libstdc++ debugging

- [x] *The Jira issue number for this PR is: MDEV-33974*
## Description
On GCC 10 or later, let us enable `_GLIBCXX_DEBUG` as well as `_GLIBCXX_ASSERTIONS` which have an impact on the GNU `libstdc++`. This should catch bugs like the one that commit 455a15fd06b0f449e72e6b10b0474e4139eeeca2 fixed.

For the clang libc++ before clang-15 there was `_LIBCPP_DEBUG`, but according to llvm/llvm-project@f3966eaf869b7bdd9113ab9d5b78469eb0f5f028 and llvm/llvm-project@13ea1343231fa4ae12fe9fba4c789728465783d7 and llvm/llvm-project@ff573a42cd1f1d05508f165dc3e645a0ec17edb5 it looks like that for proper results, a specially built debug version of `libc++` would have to be used in order to enable equivalent checks.
## Release Notes
Debug builds will try to catch some C++ API misuse in GNU libstdc++ when using GCC 10 or later.
## How can this PR be tested?
`./mtr innodb.insert_into_empty` is expected to crash as long as the commit 455a15fd06b0f449e72e6b10b0474e4139eeeca2 is not included. However, for some reason it did not crash on our CI. I tested 35aa7d655e2f4a5c717529bb9ab82363ccb6f858 in the [amd64-ubuntu-2004-debug](https://buildbot.mariadb.org/#/builders/369) environment with an additional bug injection:
```diff
--- mariadb-10.11.8/storage/innobase/row/row0merge.cc	2024-04-23 13:33:06.000000000 +0000
+++ mariadb-10.11.8/storage/innobase/row/row0merge.cc.buggy	2024-04-24 10:04:06.755847419 +0000
@@ -5379,8 +5379,8 @@
       bulk_rollback_low();
       return err;
     }
-    it->second.end_bulk_insert();
   }
+    it->second.end_bulk_insert();
   return DB_SUCCESS;
 }
 
```
This would fail during the server bootstrap:
```
/usr/include/c++/10/debug/safe_iterator.h:314:
In function:
    __gnu_debug::_Safe_iterator<_Iterator, _Sequence, _Category>::pointer 
    __gnu_debug::_Safe_iterator<_Iterator, _Sequence, 
    _Category>::operator->() const [with _Iterator = 
    std::_Rb_tree_iterator<std::pair<dict_table_t* const, 
    trx_mod_table_time_t> >; _Sequence = std::__debug::map<dict_table_t*, 
    trx_mod_table_time_t, std::less<dict_table_t*>, 
    ut_allocator<std::pair<dict_table_t* const, trx_mod_table_time_t> > >; 
    _Category = std::forward_iterator_tag; 
    __gnu_debug::_Safe_iterator<_Iterator, _Sequence, _Category>::pointer = 
    std::pair<dict_table_t* const, trx_mod_table_time_t>*]

Error: attempt to dereference a past-the-end iterator.

Objects involved in the operation:
    iterator "this" @ 0x0x7ffdb1ef83f0 {
      type = std::_Rb_tree_iterator<std::pair<dict_table_t* const, trx_mod_table_time_t> > (mutable iterator);
      state = past-the-end;
      references sequence with type 'std::__debug::map<dict_table_t*, trx_mod_table_time_t, std::less<dict_table_t*>, ut_allocator<std::pair<dict_table_t* const, trx_mod_table_time_t>, true> >' @ 0x0x7f1508f6e6b0
    }
```
## Basing the PR against the correct MariaDB version

This has been tested on 10.11, but the final version will target the 10.5 branch and not the earliest maintained branch 10.4 because it is so close to end-of-life.
## PR quality check
